Allow benchmarks to run to completion on macOS

jdm opened the pull request:
These commits fix a series of panics that I encountered while attempting to run benchmarks on macOS. The benchmarks now complete, as can be observed in #199.

highfive commented:
<img src="http://www.joshmatthews.net/warning.svg" alt="warning" height=20> **Warning** <img src="http://www.joshmatthews.net/warning.svg" alt="warning" height=20>

* These commits modify **unsafe code**. Please review it carefully!

antrik commented on src/test.rs:
```diff
@@ -428,3 +428,11 @@ fn test_reentrant() {
 }
+
+#[test]
+fn transfer_closed_sender() {
```
This is a bit nit-picky: but would you mind adding the test in a separate commit after the fix?

When making this sort of decisions, I like thinking in terms of both bisectability and revertability. Bisectability tells us that the test needs to go either in the same commit as the fix, or after it, since otherwise we would have a known-broken state in the history. Revertability says that it's better to have the test separate (either before or after the fix), since the test is probably valid regardless of the fix -- so if someone ever tries to back out the fix for any reason, the test should be still there and fail, thus making it clear that the issue needs to be addressed in some way... Considering both bisectability and revertability, we can conclude that a separate commit after the fix is indeed the best variant.

(Just to be clear, there are of course other situations as well: for tests that check newly added functionality, the same analysis would conclude that the tests need to go into the same commit as the change adding the functionality in question; and for tests that guard against regressions in things that should work even before the change, the test should go into a separate commit before the change...)

jdm commented:
I've addressed all of the comments. MachError::from can't easily handle the kernel errors because kernel_return_t and mach_msg_return_t are just aliases of the same type, and the range of errors that MachError::from handles overlaps with the range of the kernel return values.

antrik commented:
I know about the type aliasing -- but I don't understand why it is a problem? Isn't the whole idea behind this to allow uniform handling of all the errors?...

Doesn't really matter though, since I like the solution you implemented :-)

All the changes look very good to me. However, as I have limited experience with the `macos` back-end, I wonder whether it would be fine for me to r+ this myself; or better have someone else take another look?...

pcwalton commented:
Looks good to me.

jdm commented:
@bors-servo r=pcwalton

bors-servo commented:
:pushpin: Commit 0832082 has been approved by `pcwalton`

<!-- @bors-servo r=pcwalton 08320828a532116d76f06557ac5de616dae6c979 -->

bors-servo commented:
:hourglass: Testing commit 08320828a532116d76f06557ac5de616dae6c979 with merge b225f6d1c03c32963139ca43ac10a19c866a642b...

bors-servo commented:
:sunny: Test successful - [status-appveyor](https://ci.appveyor.com/project/servo/ipc-channel/build/1.0.501), [status-travis](https://travis-ci.org/servo/ipc-channel/builds/388448542?utm_source=github_status&utm_medium=notification)
Approved by: pcwalton
Pushing b225f6d1c03c32963139ca43ac10a19c866a642b to master...
